TelemetryDesignDoc.md

```diff
@@ -0,0 +1,327 @@
+  actuator: boolean
+}
+```
+
+Dashboard widgets require slightly more involved handling, which is discussed below:
+
+### Reserved metadata topic for dashboard layout definition
+
+Shuffleboard defines tabs, layouts, and widgets in a set of metadata tables, which themselves are ordinary NT entries, and the tab/layout/widget hierarchy is represented by the pathstrings used for the entry keys.  These entries live under a reserved `metadata` content root.
+
+With the addition of an explicit JSON metadata layer to NT4, we can replaec this entirely with a *single* JSON document that defines the structure of the dashboard, located under the `dashboard` topic.  This might look something like so:
+
+```typescript
+{
+  tabs: [
+    {
+      name: 'drive',
+      elements: [
+        {
+          // display name
+          name: 'Heading',
+          // widget type
+          type: 'gyro',
+          // we can point to the widget data by NT4 topic pathstring
+          data: 'drive/heading'
+        },
+        {
+          name: 'Wheel Velocities',
+          type: 'graph',
+          // multi-value plots easily represented in JSON with array type
+          data: ['drive/leftVelocity', 'drive/rightVelocity'],
+          // we can easily spec for display parameters per widget type
+          graph: {
+            limits: {
+              x: [0, 10],
+              y: [-50, 50]
+            }
+          }
+        }
+      ]
+    }
+  ]
+}
+```
+
+A simple builder API, similar to the current Shuffleboard API, can allow dashboard configuration from robot code:
+
+```java
+var exampleTab = Telemetry.getDashboardTab("Drive");
+exampleTab.addWidget(new Widgets.Gyro("Heading", "drive/heading"))
+  // Graph widget ctor can be variadic to handle multiple data sources
+  .addWidget(new Widgets.Graph("Wheel Speeds", "drive/leftVelocity", "drive/rightVelocity")
+    .withXLimits(0, 10)
+    .withYLimits(-50, 50));
+```
+
+The above differs from the Shuffleboard API in that it has direct programmatic representation for widgets in the form of the widget classes (e.g. `Widget.Gyro`).  Shuffleboard only exposes a generic string-valued `withProperties` decorator, which does not provide type safety for user-specified widget properties.
+
+### Deprecate all existing SmartDashboard and Shuffleboard methods, replace with implementation-neutral equivalents
+
+With an implementation-neutral metadata spec available, telemetry publishing is no longer tightly coupled to the user's choice of dashboard.  Accordingly, there is no need to maintain separate `SmartDashboard` and `Shuffleboard` APIs.  In fact, it is not necessary for the API shape itself to contain any reference to the dashboard implementation or spec - this will be done only where it is convenient and reduces boilerplate in user code.
+
+The unified replacement API will look something like:
+
+```java
+// Simple imperative logging directly delegates to NT4
+Telemetry.publishDouble("key", value);
```

Comment:
Does the Telemetry class keep an internal map of keys to publishers?  Note this does require a string name lookup on every call.

Reply:
The Telemetry class will need to keep multiple maps, one for each data type, to map from e.g. "doublekey" to `DoublePublisher`, "stringkey" to `StringPublisher`, etc.  One question will be how should the Telemetry class handle publishing conflicting types to the same key (e.g. `publishDouble("key")` followed by `publishString("key")`?

The NT4 API design avoids string lookups because it's handle based:
```
DoublePublisher dblPub = table.getDoubleTopic("key").publish();
...
dblPub.set(value);
...
dblPub.close();  // to stop publishing
```

Reply:
String lookups are unavoidable in the case of a synchronous write, but for a publisher bind I think we can get rid of them entirely by just mapping the supplier function to the publisher directly